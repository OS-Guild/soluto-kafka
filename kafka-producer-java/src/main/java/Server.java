import com.google.common.base.Charsets;
import com.google.common.io.CharStreams;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpServer;
import io.prometheus.client.CollectorRegistry;
import io.prometheus.client.exporter.HTTPServer;
import io.prometheus.client.hotspot.DefaultExports;
import java.io.InputStreamReader;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.util.Iterator;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;
import org.apache.kafka.common.header.Header;
import org.apache.kafka.common.header.internals.RecordHeaders;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class Server {
    Config config;
    Monitor monitor;
    HttpServer server;
    AbstractProducer producer;

    Server(Config config, Monitor monitor, AbstractProducer producer) {
        this.config = config;
        this.monitor = monitor;
        this.producer = producer;
    }

    public Server start() throws IOException {
        server = HttpServer.create(new InetSocketAddress(Config.PORT), 0);
        isAliveGetRoute(server);
        producePostRoute(server);
        if (Config.USE_PROMETHEUS) {
            DefaultExports.initialize();
            new HTTPServer(server, CollectorRegistry.defaultRegistry, false);
        } else {
            server.start();
        }
        return this;
    }

    public void close() {
        server.stop(0);
    }

    private void isAliveGetRoute(HttpServer server) {
        var httpContext = server.createContext("/isAlive");
        httpContext.setHandler(
            new HttpHandler() {

                @Override
                public void handle(HttpExchange exchange) throws IOException {
                    if (!exchange.getRequestMethod().equals("GET")) {
                        exchange.sendResponseHeaders(404, -1);
                        return;
                    }
                    if (!producer.ready()) {
                        exchange.sendResponseHeaders(500, -1);
                        return;
                    }
                    exchange.sendResponseHeaders(204, -1);
                }
            }
        );
    }

    private void producePostRoute(HttpServer server) {
        var httpContext = server.createContext("/produce");
        httpContext.setHandler(
            new HttpHandler() {

                @Override
                public void handle(HttpExchange exchange) throws IOException {
                    if (!exchange.getRequestMethod().equals("POST")) {
                        exchange.sendResponseHeaders(404, -1);
                        return;
                    }
                    var body = CharStreams.toString(new InputStreamReader(exchange.getRequestBody(), Charsets.UTF_8));
                    try {
                        List<ProducerRequest> requests = StreamSupport
                            .stream(new JSONArray(body).spliterator(), false)
                            .map(
                                jsonItem -> {
                                    var item = new JSONObject(jsonItem.toString());
                                    return new ProducerRequest(
                                        tryGetValue(item, "topic", null),
                                        tryGetValue(item, "key", null),
                                        tryGetValue(item, "message", "value"),
                                        tryGetHeaders(item)
                                    );
                                }
                            )
                            .collect(Collectors.toList());
                        for (ProducerRequest producerRequest : requests) {
                            producer.produce(producerRequest);
                        }
                        exchange.sendResponseHeaders(204, -1);
                    } catch (Exception e) {
                        e.printStackTrace();
                        var response = e.getMessage();
                        exchange.sendResponseHeaders(
                            e instanceof IllegalArgumentException ? 400 : 500,
                            response.getBytes().length
                        );
                        var os = exchange.getResponseBody();
                        os.write(response.getBytes());
                        os.close();
                    }
                }
            }
        );
    }

    private Iterable<Header> tryGetHeaders(JSONObject item) {
        if (item.has("headers")) {
            JSONObject headersJson = item.getJSONObject("headers");
            Iterator<String> keys = headersJson.keys();
            if (keys.hasNext()) {
                RecordHeaders headers = new RecordHeaders();
                while (keys.hasNext()) {
                    String key = keys.next();
                    if (headersJson.has(key)) {
                        try {
                            headers.add(key, headersJson.getString(key).getBytes());
                        } catch (JSONException e) {
                            headers.add(key, null);
                        }
                    }
                }
                return headers;
            }
        }
        return null;
    }

    private static String tryGetValue(JSONObject json, String option1, String option2) {
        if (json.has(option1)) {
            return json.get(option1).toString();
        }
        if (json.has(option2)) {
            return json.get(option2).toString();
        }

        if (option2 == null) {
            throw new IllegalArgumentException(option1 + " is missing");
        }
        throw new IllegalArgumentException(option2 + " is missing");
    }
}
