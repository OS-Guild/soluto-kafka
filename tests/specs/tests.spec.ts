import delay from 'delay';
import fetch from 'node-fetch';
import Server from 'simple-fake-server-server-client';
import {range} from 'lodash';

import checkReadiness from '../checkReadiness';
import * as uuid from 'uuid';
import {PORTS} from '../fakeGRPCServer/constants';

jest.setTimeout(180000);

const fakeHttpServer = new Server({
    baseUrl: `http://localhost`,
    port: 3000,
});

describe('tests', () => {
    beforeAll(async () => {
        await expect(checkReadiness(['foo', 'bar', 'retry', 'dead-letter', 'unexpected'])).resolves.toBeTruthy();
    });

    beforeEach(async () => {
        await fakeHttpServer.clear();
    });

    it('liveliness', async () => {
        await delay(1000);
        const producer = await fetch('http://localhost:6000/isAlive');
        const consumer = await fetch('http://localhost:4001/isAlive');
        expect(producer.ok).toBeTruthy();
        expect(consumer.ok).toBeTruthy();
    });

    it('should produce and consume', async () => {
        const callId = await mockHttpTarget('/consume', 200);

        await produce('http://localhost:6000/produce', [
            {
                topic: 'foo',
                key: 'thekey',
                value: {data: 'foo'},
                headers: {eventType: 'test1', source: 'test-service1', nullHeader: null},
            },
        ]);
        await delay(1000);
        await produce('http://localhost:6000/produce', [
            {
                topic: 'bar',
                key: 'thekey',
                value: {data: 'bar'},
                headers: {eventType: 'test2', source: 'test-service2'},
            },
        ]);
        await delay(1000);

        const {hasBeenMade, madeCalls} = await fakeHttpServer.getCall(callId);
        expect(hasBeenMade).toBeTruthy();
        expect(madeCalls.length).toBe(2);
        const actualHeaders1 = JSON.parse(madeCalls[0].headers['x-record-headers']);
        const actualHeaders2 = JSON.parse(madeCalls[1].headers['x-record-headers']);
        expect(madeCalls[0].headers['x-record-topic']).toBe('foo');
        expect(actualHeaders1!.eventType).toEqual('test1');
        expect(actualHeaders1!.source).toEqual('test-service1');
        expect(actualHeaders1!.nullHeader).toEqual(null);
        expect(madeCalls[1].headers['x-record-topic']).toBe('bar');
        expect(actualHeaders2!.eventType).toEqual('test2');
        expect(actualHeaders2!.source).toEqual('test-service2');
        expect(parseInt(madeCalls[0].headers['x-record-timestamp'])).not.toBe(NaN);
        expect(parseInt(madeCalls[1].headers['x-record-timestamp'])).not.toBe(NaN);
    });

    it('should consume bursts of records', async () => {
        const callId = await mockHttpTarget('/consume', 200);

        const recordsCount = 1000;
        const records = range(recordsCount).map(() => ({topic: 'foo', key: uuid(), value: {data: 'foo'}}));

        await produce('http://localhost:6000/produce', records);
        await delay(recordsCount * 10);

        const {madeCalls} = await fakeHttpServer.getCall(callId);
        expect(madeCalls.length).toBe(recordsCount);
    });

    it('consumer should produce to dead letter topic when target response is 400', async () => {
        await mockHttpTarget('/consume', 400);
        const callId = await mockHttpTarget('/deadLetter', 200);

        await produce('http://localhost:6000/produce', [{topic: 'foo', key: uuid(), value: {data: 'foo'}}]);
        await delay(5000);

        const {hasBeenMade, madeCalls} = await fakeHttpServer.getCall(callId);
        expect(hasBeenMade).toBeTruthy();
        expect(madeCalls[0].headers['x-record-original-topic']).toEqual('foo');
    });

    it('consumer should produce to retry topic when target response is 500', async () => {
        await mockHttpTarget('/consume', 500);
        const callId = await mockHttpTarget('/retry', 200);

        await produce('http://localhost:6000/produce', [{topic: 'foo', key: uuid(), value: {data: 'foo'}}]);
        await delay(5000);

        const {hasBeenMade, madeCalls} = await fakeHttpServer.getCall(callId);
        expect(hasBeenMade).toBeTruthy();
        expect(madeCalls[0].headers['x-record-original-topic']).toEqual('foo');
    });

    it('consumer should terminate on an unexpected error', async () => {
        await delay(1000);
        let consumerLiveliness = await fetch('http://localhost:4002/isAlive');
        expect(consumerLiveliness.ok).toBeTruthy();

        await produce('http://localhost:6000/produce', [
            {topic: 'unexpected', key: uuid(), value: {data: 'unexpected'}},
        ]);
        await delay(10000);

        consumerLiveliness = await fetch('http://localhost:4002/isAlive');
        expect(consumerLiveliness.ok).toBeFalsy();
    });

    it('producer request validation', async () => {
        const method = 'post';
        const producerUrl = 'http://localhost:6000/produce';
        const headers = {'Content-Type': 'application/json'};
        let response;

        response = await fetch(producerUrl, {
            method,
            body: JSON.stringify([{key: 'key', value: {data: 1}}]),
            headers,
        });
        expect(response.status).toBe(400);
        expect(await response.text()).toBe('topic is missing');

        response = await fetch(producerUrl, {
            method,
            body: JSON.stringify([{topic: 'bar', value: {data: 1}}]),
            headers,
        });
        expect(response.status).toBe(400);
        expect(await response.text()).toBe('key is missing');

        response = await fetch(producerUrl, {
            method,
            body: JSON.stringify([{topic: 'bar', key: 'key'}]),
            headers,
        });
        expect(response.status).toBe(400);
        expect(await response.text()).toBe('value is missing');
    });
});

describe('gRPC tests', () => {
    const FAKE_GRPC_SERVER_URL = `http://localhost:${PORTS.express}`;

    beforeEach(async () => {
        await fetch(`${FAKE_GRPC_SERVER_URL}/clear`);
    });

    it('should pass all data and headers', async () => {
        const topic = 'bar';
        const payload = {data: 'bar'};

        await produce('http://localhost:6000/produce', [
            {
                topic,
                key: 'thekey',
                value: payload,
            },
        ]);

        await delay(1000);

        const fakeGRPCResponse = await fetch(`${FAKE_GRPC_SERVER_URL}/getCallHistory`);
        const {
            calls: [call],
        } = await fakeGRPCResponse.json();

        expect(call?.headers?.recordOffset).toBeDefined();
        expect(typeof call?.headers?.recordOffset).toBe('number');
        expect(call?.headers?.recordTimestamp).toBeDefined();
        expect(typeof call?.headers?.recordTimestamp).toBe('number');
        expect(call?.headers?.topic).toBe(topic);
        expect(call?.payload).toMatchObject(payload);
    });
});

describe.only('message order tests', () => {
    const FAKE_GRPC_SERVER_URL = `http://localhost:${PORTS.express}`;

    beforeEach(async () => {
        await fetch(`${FAKE_GRPC_SERVER_URL}/clear`);
    });

    it('should receive calls in same order they were produced', async () => {
        const topic = 'foo';
        const payload = {data: 'bar'};
        const numberOfMessages = 10;
        const delayMillis = 3000;
        let index = 0;

        while (index < numberOfMessages) {
            await produce('http://localhost:6000/produce', [
                {
                    topic,
                    key: 'samePartitioningKey',
                    headers: {
                        executionDelay: index % 2 > 0 ? null : `${delayMillis}`,
                    },
                    value: {
                        index,
                        ...payload,
                    },
                },
            ]);
            index++;
        }

        await delay(numberOfMessages * delayMillis);

        const fakeGRPCResponse = await fetch(`${FAKE_GRPC_SERVER_URL}/getCallHistory`);
        const {calls} = await fakeGRPCResponse.json();
        expect(calls.length).toBe(numberOfMessages);
        index = 0;
        while (index < numberOfMessages) {
            expect(calls[index].payload.index).toBe(index);
            index++;
        }
    });
});

const produce = (url: string, batch: any[]) =>
    fetch(url, {
        method: 'post',
        body: JSON.stringify(batch),
        headers: {'Content-Type': 'application/json'},
    });

const mockHttpTarget = (route: string, statusCode: number) =>
    fakeHttpServer.mock({
        method: 'post',
        url: route,
        statusCode,
    });
