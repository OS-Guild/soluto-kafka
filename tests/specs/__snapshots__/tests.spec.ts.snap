// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests should produce and consume 1`] = `"{\\"eventType\\":\\"test1\\",\\"source\\":\\"test-service1\\"}"`;

exports[`tests should produce and consume 2`] = `"{\\"eventType\\":\\"test2\\",\\"source\\":\\"test-service2\\"}"`;
